### ISSUE TITLE
Does not compile for ARM

### ISSUE THREAD

**Issue by mattchatporter** [lh3/bwa#300]
Attempting to run `make` on an ARM system yields the following error:

```
ksw.c:29:10: fatal error: emmintrin.h: No such file or directory
29 | #include <emmintrin.h>
```

**Comment by tomaskopsa** [comment]
bwa requires SSE2 instruction set. emmintrin.h is header file for SSE2 intrinsics.
You are compiling the code for platform without SSE2 support.

**Comment by mp-consulting** [comment]
Exactly - and considering that the following are true:

* ARM workloads are typically 40% more cost-effective when compared to running the same workload on an Intel CPU (e.g. EC2 c6g)
* Alignment is one of the most expensive steps in whole genome sequencing analysis
* Compiling software for ARM can often be as simple as using the appropriate ARM compiler and setting the appropriate flags

It makes good sense then for a high-value, (hopefully) low-effort update to be made to this project to support ARM. Maybe there's more work involved here than a simple compiler & flag switch, but if another existing aligner adds ARM support while BWA has no announced plans to support this, concerns around cost performance as well as future-looking hardware support will cause it to start looking less attractive as a real-world use-case tool.

**Comment by lh3** [comment]
Where does the 40% number come from? Even if that is true for a particular benchmark, it is probably not generalizable. You are underestimating how high-performance tools are developed these days. At times it is more complicated than changing flags. In case of bwa-mem, you need to emulate SSE2 with simde or sse2neon. This is not a one-to-one translation and will cost performance. It is even trickier with bwa-mem as most ARM CPUs don't support 256-bit SIMD, let alone 512-bit. Also, the cache performance/bandwidth, branch predictions etc are different between CPUs. You need architecture specific optimization. I doubt simple port can match the performance on x86 CPUs.

The ARM ecosystem is tiny right now. I will see how it evolves in particular with the availability of ARM Mac, but for now, no plan for ARM support.

**Comment by barkoneus** [comment]
Good news is that binaries compiled for Intel Macs run on M1 Macs, confirmed as I recently upgraded my 2018 Intel MacBook Pro to a new M1 MacBook Pro.

Emulated bwa on my M1 laptop is about 25% faster than native bwa on the Intel laptop which was a pleasant surprise.

**Comment by khanshahan** [comment]
> Good news is that binaries compiled for Intel Macs run on M1 Macs, confirmed as I recently upgraded my 2018 Intel MacBook Pro to a new M1 MacBook Pro.
> 
> Emulated bwa on my M1 laptop is about 25% faster than native bwa on the Intel laptop which was a pleasant surprise.

Hey Hi , can you please tell me how can I achieve the same? I have intelbased Macbook while my sister has M1. How can I copy the bwa from my machine to her system?

**Comment by barkoneus** [comment]
Yes, all you need to do is copy over the bwa executable file from the Intel machine to the M1 machine. Nothing else needs to be done.

**Comment by khanshahan** [comment]
Thank you very much @barkoneus , this worked :)

**Comment by markotitel** [comment]
This can help to compile for ARM
https://aws.amazon.com/blogs/publicsector/generalized-approach-benchmarking-genomics-workloads-cloud-bwa-read-aligner-graviton2/

**Comment by edubielalpizar** [comment]
> 

sounds easy but HOW to do it?  (1) where is the executable (2) where (which directory) to copy from intel into the new M1 ?

:(

**Comment by martin-g** [comment]
It would be awesome if the support for ARM64 is added!

**Comment by V-Z** [comment]
BTW, the issue seems same for PPC64, S390X and RISC-V architectures.

**Comment by julien-faye** [comment]
How can we help to get the support for ARM64 ?

**Comment by jmarshall** [comment]
See also PR #359.
